presto-main/src/main/java/com/facebook/presto/resourcemanager/ClusterQueryTrackerService.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.resourcemanager;
+
+import com.facebook.drift.client.DriftClient;
+import com.facebook.presto.util.PeriodicTaskExecutor;
+
+import javax.annotation.PostConstruct;
+import javax.annotation.PreDestroy;
+import javax.inject.Inject;
+
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static java.util.Objects.requireNonNull;
+
+public class ClusterQueryTrackerService
+{
+    private final DriftClient<ResourceManagerClient> resourceManagerClient;
+    private final ScheduledExecutorService executorService;
+    private final long runningTaskCountFetchIntervalMillis;
+    private AtomicReference<Integer> runningTaskCount;
+    private final PeriodicTaskExecutor runningTaskCountUpdater;
+
+    @Inject
+    public ClusterQueryTrackerService(
+            @ForResourceManager DriftClient<ResourceManagerClient> resourceManagerClient,
+            @ForResourceManager ScheduledExecutorService executorService,
+            ResourceManagerConfig resourceManagerConfig)
+    {
+        this.resourceManagerClient = requireNonNull(resourceManagerClient, "resourceManagerClient is null");
+        this.executorService = requireNonNull(executorService, "executorService is null");
+        this.runningTaskCountFetchIntervalMillis = requireNonNull(resourceManagerConfig, "resourceManagerConfig is null").getRunningTaskCountFetchInterval().toMillis();
+        this.runningTaskCountUpdater = new PeriodicTaskExecutor(runningTaskCountFetchIntervalMillis, executorService, () -> updateRunningTaskCount());
+    }
+
+    @PostConstruct
+    public void init()
+    {
+        runningTaskCountUpdater.start();
+    }
+
+    @PreDestroy
+    public void stop()
+    {
+        runningTaskCountUpdater.stop();
+    }
+
+    public int getRunningTaskCount()
+    {
+        return runningTaskCount.get();
+    }
+
+    private void updateRunningTaskCount()
+    {
+        this.runningTaskCount.set(resourceManagerClient.get().getRunningTaskCount());
```

Comment:
do we have multiple threads updating runningTaskCount?. Curious why an `AtomicReference` is used?.  I could see only a scheduled thread updating this. 

If we do have multiple threads, why not use an `AtomicInteger` instead?

Reply:
might be able to get away with a volatile since only visibility is in question here. The var is only written from one thread always and read from some other thread.